src/TestUtils/src/DeviceTests/AssertionExtensions.Windows.cs

```diff
@@ -165,99 +165,98 @@ public static async Task<T> AttachAndRun<T>(this FrameworkElement view, Func<Win
 			if (view.Parent is Border wrapper)
 				view = wrapper;
 
-			TaskCompletionSource? tcs = null;
-			TaskCompletionSource? unloadedTcs = null;
-
-			if (view.Parent == null)
+			// If the view has a parent, it's already attached to the UI
+			if (view.Parent != null)
 			{
-				T result;
+				// Window is not a XAML type so is never on the hierarchy
+				var window = (Window)mauiContext!.Services!.GetService(typeof(Window))!;
 
-				try
-				{
-					await _attachAndRunSemaphore.WaitAsync();
+				if (window.Content.XamlRoot != view.XamlRoot)
+					throw new Exception("The window retrieved from the service is different than the window this view is attached to");
 
-					// prepare to wait for element to be in the UI
-					tcs = new TaskCompletionSource();
-					unloadedTcs = new TaskCompletionSource();
+				return await Run(() => action(window));
+			}
 
-					view.Loaded += OnViewLoaded;
+			// If the view has no parent, we need to attach it to the UI by creating a new window and using that as the host
+			try
+			{
+				await _attachAndRunSemaphore.WaitAsync();
 
-					// attach to the UI
-					Grid grid;
-					var window = (Window)mauiContext!.Services!.GetService(typeof(Window))!;
+				// prepare to wait for element to be in the UI
+				var loadedTcs = new TaskCompletionSource();
+				var unloadedTcs = new TaskCompletionSource();
 
-					if (window.Content is not null)
-						throw new Exception("The window retrieved from the service is already attached to existing content");
+				view.Loaded += OnViewLoaded;
 
-					window.Content = new Grid
+				// attach to the UI
+				Grid viewContainer;
+				var window = (Window)mauiContext!.Services!.GetService(typeof(Window))!;
+
+				if (window.Content is not null)
+					throw new Exception("The window retrieved from the service is already attached to existing content");
+
+				window.Content = new Grid
```

Comment:
Do we want this to be a Grid? Would it make sense for this to be a ContentPanel or a a custom panel? That way we can call into the IView.Measure and IView.Arrange calls?

Which should let us remove calls like this 
https://github.com/dotnet/maui/blob/main/src/Controls/tests/DeviceTests/Elements/Frame/FrameTests.cs#L438-L440

Which are now happening outside the natural layout cycle

Reply:
I think I am using a double Grid because the first grid is the root, and then the second one keeps the view at the size it wants and centers. Maybe the inner grid should be the ContentPanel?